pkg/ebpf/c/tcp-queue-length-kern.c

```diff
@@ -0,0 +1,136 @@
+#include <linux/kconfig.h>
+#define KBUILD_MODNAME "foo"
+#include <linux/ptrace.h>
+#include <linux/bpf.h>
+#include <net/inet_sock.h>
+#include <linux/tcp.h>
+
+#include "tcp-queue-length-kern-user.h"
+
+BPF_HASH(queue, struct sock *, struct stats);
+
+BPF_HASH(who_recvmsg, u64, struct sock *);
+BPF_HASH(who_sendmsg, u64, struct sock *);
+
+// TODO: replace all `bpf_probe_read` by `bpf_probe_read_kernel` once we can assume that we have at least kernel 5.5
+static inline int check_sock(struct sock *sk) {
+  struct stats zero = {
+    .rqueue = {
+      .min = 2^32-1,
+      .max = 0
+    },
+    .wqueue = {
+      .min = 2^32-1,
+      .max = 0
+    }
+  };
+
+  struct stats *s = queue.lookup_or_init(&sk, &zero);
+  if (s == NULL) return 0;
+
+  /*
+   * We assume here that only one thread will read and/or write to a given socket.
+   * Indeed, having several unsynchronized threads attempting to read and/or write to a socket
+   * would corrupt the stream.
+   * If that assumption was wrong, we would need to make the following piece of code thread safe.
+   * In that case, per-cpu hash would be a better solution than mutex.
+   */
+  if (s->pid == 0) {
+    s->pid = bpf_get_current_pid_tgid() >> 32;
+
+    struct task_struct *cur_tsk = (struct task_struct *)bpf_get_current_task();
+    struct css_set *css_set;
+    if (!bpf_probe_read(&css_set, sizeof(css_set), &cur_tsk->cgroups)) {
+      struct cgroup_subsys_state *css;
+      // TODO: Do not arbitrarily pick the first subsystem
+      if (!bpf_probe_read(&css, sizeof(css), &css_set->subsys[0])) {
+        struct cgroup *cgrp;
+        if (!bpf_probe_read(&cgrp, sizeof(cgrp), &css->cgroup)) {
+          struct kernfs_node *kn;
+          if (!bpf_probe_read(&kn, sizeof(kn), &cgrp->kn)) {
+            const char *name;
+            if (!bpf_probe_read(&name, sizeof(name), &kn->name)) {
+              bpf_probe_read_str(&s->cgroup_name, sizeof(s->cgroup_name), name);
+            }
+          }
+        }
+      }
+    }
+
+    const struct inet_sock *ip = inet_sk(sk);
+    bpf_probe_read(&s->conn.saddr, sizeof(s->conn.saddr), &ip->inet_saddr);
+    bpf_probe_read(&s->conn.daddr, sizeof(s->conn.daddr), &ip->inet_daddr);
+    bpf_probe_read(&s->conn.sport, sizeof(s->conn.sport), &ip->inet_sport);
+    bpf_probe_read(&s->conn.dport, sizeof(s->conn.dport), &ip->inet_dport);
+  }
+
+  const struct tcp_sock *tp = tcp_sk(sk);
+
+  u32 rcv_nxt, copied_seq, write_seq, snd_una;
+  bpf_probe_read(&rcv_nxt,    sizeof(rcv_nxt),    &tp->rcv_nxt   );  // What we want to receive next
+  bpf_probe_read(&copied_seq, sizeof(copied_seq), &tp->copied_seq);  // Head of yet unread data
+  bpf_probe_read(&write_seq,  sizeof(write_seq),  &tp->write_seq );  // Tail(+1) of data held in tcp send buffer
+  bpf_probe_read(&snd_una,    sizeof(snd_una),    &tp->snd_una   );  // First byte we want an ack for
+
+  int rqueue = rcv_nxt - copied_seq;
+  if (rqueue < 0) rqueue = 0;
```

Comment:
when does this happen?

Reply:
This can happen if the process reading the data from the socket is running on a CPU concurrently to data being received from the NIC on another CPU:
```
     +------+               +------+
     | CPU1 |               | CPU2 |
     +------+               +------+
        |                      |
        |              start of IRQ to process
        |              incoming network packet
        |                      |
        |              put received data into socket buffer
        |                      |
user-land program reads        |
data from the socket           |
        |                      |
the `read` syscall             |
updates `copied_seq`           |
        |<--------------------------------------- Here is where the kretprobe__tcp_recvmsg probe is invoked
        |                      |                  The eBPF ret-probe is invoked once the data has been read.
        |              net RX process updates     At this point, `copied_seq` has been updated.
        |              `rcv_nxt`                  But potentially, `rcv_nxt` hasn’t been updated by CPU2 yet.
        |                      |
```

See https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=49d09007879ce7bee36ab453c73e97c00adce884 for further details.
That’s the kernel commit that introduced this check. (See the adding of `max_t` in `tcp_diag_get_info` which is the kernel function that computes the size of send/receive buffer like what this function does)